synapse/rest/admin/federation.py

```diff
@@ -16,6 +16,7 @@
 from typing import TYPE_CHECKING, Tuple
 
 from synapse.api.errors import Codes, NotFoundError, SynapseError
+from synapse.federation.transport.server._base import Authenticator
```

Comment:
I got the error
```
builtins.ImportError: cannot import name 'Authenticator' from partially initialized module 'synapse.federation.transport.server._base' (most likely due to a circular import) (/synapse/build/synapse/federation/transport/server/_base.py)
```

I do not know how is the best way to solve this.
One option is do not use `_reset_retry_timings`
https://github.com/matrix-org/synapse/blob/66d7aa783a8ad1b3eb6d17ba086e2dda8a912ab6/synapse/federation/transport/server/_base.py#L121-L137

and call manually
```python
store.set_destination_retry_timings(origin, None, 0, 0)
notifier.notify_remote_server_up(origin)
```

Reply:
Looks like you got this working, but it's a bit of a shame to have to duplicate those two lines. Was there any additional context with the import error?

Reply:
Aha:

```
2021-12-23T16:24:01.3115062Z Traceback (most recent call last):
2021-12-23T16:24:01.3117986Z   File "/home/runner/work/synapse/synapse/.tox/py-noextras/lib/python3.10/site-packages/twisted/trial/runner.py", line 693, in loadByName
2021-12-23T16:24:01.3119093Z     return self.suiteFactory([self.findByName(name, recurse=recurse)])
2021-12-23T16:24:01.3120526Z   File "/home/runner/work/synapse/synapse/.tox/py-noextras/lib/python3.10/site-packages/twisted/trial/runner.py", line 453, in findByName
2021-12-23T16:24:01.3121458Z     obj = reflect.namedModule(searchName)
2021-12-23T16:24:01.3122925Z   File "/home/runner/work/synapse/synapse/.tox/py-noextras/lib/python3.10/site-packages/twisted/python/reflect.py", line 157, in namedModule
2021-12-23T16:24:01.3123764Z     topLevel = __import__(name)
2021-12-23T16:24:01.3124407Z   File "/home/runner/work/synapse/synapse/tests/api/test_auth.py", line 32, in <module>
2021-12-23T16:24:01.3125015Z     from tests import unittest
2021-12-23T16:24:01.3125657Z   File "/home/runner/work/synapse/synapse/tests/unittest.py", line 53, in <module>
2021-12-23T16:24:01.3126542Z     from synapse.federation.transport import server as federation_server
2021-12-23T16:24:01.3127524Z   File "/home/runner/work/synapse/synapse/synapse/federation/transport/server/__init__.py", line 21, in <module>
2021-12-23T16:24:01.3128527Z     from synapse.federation.transport.server._base import (
2021-12-23T16:24:01.3129664Z   File "/home/runner/work/synapse/synapse/synapse/federation/transport/server/_base.py", line 32, in <module>
2021-12-23T16:24:01.3130439Z     from synapse.server import HomeServer
2021-12-23T16:24:01.3131161Z   File "/home/runner/work/synapse/synapse/synapse/server.py", line 52, in <module>
2021-12-23T16:24:01.3131941Z     from synapse.events.spamcheck import SpamChecker
2021-12-23T16:24:01.3132802Z   File "/home/runner/work/synapse/synapse/synapse/events/spamcheck.py", line 31, in <module>
2021-12-23T16:24:01.3133605Z     from synapse.rest.media.v1._base import FileInfo
2021-12-23T16:24:01.3134368Z   File "/home/runner/work/synapse/synapse/synapse/rest/__init__.py", line 18, in <module>
2021-12-23T16:24:01.3255435Z     from synapse.rest import admin
2021-12-23T16:24:01.3256406Z   File "/home/runner/work/synapse/synapse/synapse/rest/admin/__init__.py", line 43, in <module>
2021-12-23T16:24:01.3257226Z     from synapse.rest.admin.federation import (
2021-12-23T16:24:01.3258116Z   File "/home/runner/work/synapse/synapse/synapse/rest/admin/federation.py", line 19, in <module>
2021-12-23T16:24:01.3259192Z     from synapse.federation.transport.server._base import Authenticator
2021-12-23T16:24:01.3261751Z builtins.ImportError: cannot import name 'Authenticator' from partially initialized module 'synapse.federation.transport.server._base' (most likely due to a circular import) 
```

To import `_base`, we have to import `Homeserver`. To do that, we have to import `synapse.rest`. Which with this change, imports `Authenticator` from `_base`: a loop.

We could hack around this by e.g. only importing `Homeserver` in `_base` if `typing.TYPE_CHECKING` and changing the type annotations to be strings. Not convinced that's the right way to do it---would be interested in thoughts from @matrix-org/synapse-core here

Reply:
it looks like @dklimpel has worked around this anyway now, but:

1. The `_` in `_base` denotes it as a "private" module -  ie, it is only for use within `synapse.federation.transport.server, so it should not be imported into `synapse.rest.admin` anyway. (If you need to reference `Authenticator`, you can `from synapse.federation.transport.server import Authenticator`, but you'll still have a loop here).
2. Yes, `Homeserver` should probably only be imported if `typing.TYPE_CHECKING` - that's what most of the codebase does, for much the same reason.

`synapse.rest` is pretty high-level, so I think importing from `synapse.federation.transport.server` is probably ok.